Monitoring and time changes (#61)
ietf-wg-scone/scone: draft-ietf-scone-protocol.md

{: #ex-rates title="Examples of SCONE signals and corresponding rates"}

## Monitoring Period {#time}

The time over which throughput advice applies is defined to be

> a period of 67 seconds.

Reviewer: Yes, I think that needs some explanation.

{: #ex-rates title="Examples of SCONE signals and corresponding rates"}

## Monitoring Period {#time}

The time over which throughput advice applies is defined to be
a period of 67 seconds.

This value aims to strike a balance between extremes.
Longer periods allow applications more flexibility
in terms of how to allocate bandwidth over time.
Shorter periods allow networks to administer policies more tightly.
Also, when circumstances change,
applications are better able recover
without exceeding limits for a significant time
if they have exceeded limits.

The choice of 67 seconds, as a prime number,
also ensures that problems in tuning applications
are not synchronized with other periodic effects.
Any repeating phenomenon at this interval is therefore
unlikely to be due to other periodic effects.

> 

Author: I don't like saying "we did this because that was the consensus outcome".  I prefer to use the reasons we used to decide.

Your point about seconds vs. milliseconds is well-taken.  I'll include the video segment length/key frame interval as an example.

> 

Reviewer 2: The new text is improved, thanks.

{: #ex-rates title="Examples of SCONE signals and corresponding rates"}

## Monitoring Period {#time}

The time over which throughput advice applies is defined to be
a period of 67 seconds.

Protocol participants can use a different period,
depending on their role.
Senders can limit their send rate over any time period
up to 67 seconds.
Network elements can monitor and apply limits to send rates
using time period of at least 67 seconds.

> The choice of 67 seconds aims is a compromise between competing interests.

Author: ```suggestion
The choice of 67 seconds is a compromise between competing interests.
```

Longer periods allow applications more flexibility
in terms of how to allocate bandwidth over time.
Shorter periods allow networks to administer policies more tightly.
Also, when circumstances change,
applications are better able recover
without exceeding limits for a significant time
if they have exceeded limits.

The choice of 67 seconds, as a prime number,

> also ensures that problems in tuning applications

Reviewer 3: Does it "ensure" this (guarantee)? I wonder if it mitigates potential problems....

Reviewer 4: That might be a bit too strong language, yes.

> The choice of 67 seconds, as a prime number,
> also ensures that problems in tuning applications
> are not synchronized with other periodic effects
> that might be measured in whole seconds.

Reviewer 4: ```suggestion
The choice of 67 seconds, as a prime number,
also helps avoid synchronization with other periodic
effects that are commonly measured in whole seconds.
```

This includes segment length or key frame intervals in video applications,
but also includes timers for middleboxes; see {{Section 4.3 of ?RFC4787}}.

> Any repeating phenomenon at this interval is therefore

Author: ```suggestion
Any repeating phenomenon at a 67 second interval is therefore
```

The time over which throughput advice applies is defined to be
a period of 67 seconds.

Protocol participants can use a different period,
depending on their role.
Senders can limit their send rate over any time period
up to 67 seconds.
Network elements can monitor and apply limits to send rates
using time period of at least 67 seconds.

The choice of 67 seconds aims is a compromise between competing interests.

> Longer periods allow applications more flexibility

Reviewer 2: Are you sure about aims in that sentence?

The choice of 67 seconds, as a prime number,
also ensures that problems in tuning applications
are not synchronized with other periodic effects
that might be measured in whole seconds.
This includes segment length or key frame intervals in video applications,
but also includes timers for middleboxes; see {{Section 4.3 of ?RFC4787}}.
Any repeating phenomenon at this interval is therefore

> unlikely to be due to other periodic effects.

Reviewer 2: The sentence in the middle is much improved, but I don't fancy much the undefined "this" in the last sentence. Matter of taste, but I would write "at a 67 seconds interval".

And just to be facetious, why not 61, or 71, or 73?

Author: There's no real reason for it to be 67 over any of those.  Also, 68 probably wouldn't be so bad either.

Once the throughput advice signal is updated,
the network element updates the UDP checksum for the datagram.

A network element needs to ensure that it sends updated rate signals
with no more than a monitoring period ({{time}}) between each update.
Because this depends on the availability of SCONE packets
and packet loss can cause signals to be missed,
network elements might need to update more often.

Senders that send a SCONE packet
or network elements that update SCONE packets

> every 20&ndash;30 seconds is likely sufficient to ensure that throughput advice is not lost.

Reviewer 5: I see, thank you for the clarification.

To me it seems that "add" in "senders can _add_ a small random delay" was the source of my confusion. Could we change that to "include"?

that respect throughput advice.
This section outlines an exemplary algorithm
for network elements.

This monitoring algorithm is guidance only.
Network elements are advised that monitoring
any more strictly than the following likely invalidates their advice.
Network elements could choose not to monitor in this way
or use a looser monitoring approach.
For instance, monitoring over a longer time window
than the monitoring period (67s)
or using a higher rate than is signaled
is possible.

To operate this algorithm,
the minimal state a network element maintains is:
* the current rate limit,
* any state necessary to monitor throughput
  (that is, throughput usage state, such as the state in {{sliding-window}}),
* a timer used for rate increases, and
* the time at which that rate limit was last updated,

When advice is set or updated,
the network element waits until
it receives the next SCONE packet on affected flows.
It then updates the throughput advice in that packet ({{apply}})
and updates its monitoring state as follows:

* If the signaled rate is the same as the current rate,
  set the last update time to the current time.

* If the signaled rate is higher than the current value,

>   then:

Reviewer 5: PS. As a high order point, I wonder the proposed algorithm is incorrect in sense that it retains old limit iff the rate limit is increased.

IMO, the correct approach _for network elements_ is to retain old limits when the rate limit is reduced, but apply the new rate limits immediately when they are increased.

This is in contrast to how _endpoints_ should behave; i.e., is retain old limits when the rate limit is increased, but apply the new limits ASAP when they are decreased.

Author: It retains the old rate if the new rate is the same or higher (not "iff").

How could you retain the old limit if the new one reduced?

Let's say that you start at 1Mbps and send at that rate for an extended time.  At the point the number drops to 200Kbps, you have just spent the last however long sending at a higher rate.  Let's assume that the send rate is approximately constant.  In that case, because the rate dropped to 1/5th of the previous, if any period you measure includes the period of period/5 seconds prior to the change in advice, you cannot send anything at all 4/5ths of a monitoring period.

I think that I'm going to need you to more clearly explain exactly how you imagine endpoints and network elements behave here, because I'm clearly not understanding your point.  I don't know how to implement:

> IMO, the correct approach for network elements is to retain old limits when the rate limit is reduced, but apply the new rate limits immediately when they are increased.

Reviewer 5: > I think that I'm going to need you to more clearly explain exactly how you imagine endpoints and network elements behave here, because I'm clearly not understanding your point.

Let's reuse the previous example.

1. [t=0] NE indicates 1Mbps
2. [t=0] endpoints start sending at 1Mbps, which is expected to continue for 50 seconds
3. [t=20] NE indicates 100Kbps

Consider this endpoint a real-time video application. By receiving a bit rate advice via a SCONE packet at t=0, it is assuming that it can send at 1Mbps. The application updates its send rate only once every 50 seconds, because 50 is smaller than 67.

As the application updates its send rate every 50 seconds, it fails to adopt to the bit rate reduction at t=20. The logic being proposed in this pull request would call out excess usage at t=26.7; i.e., 6.7 seconds after the 100Kbps signal is sent.

Note that the grace period in this example is 6.7 seconds, because the rate limit was reduced to a 1/10th of the original. As the ratio of reduction is sharper (it could theoretically be as large as 100K/199.5G), the "grace period" becomes smaller.

One might argue that the fix that we can apply is to require real-time video applications adjust to the new bit rate more promptly than 6.7 seconds, and disregard network elements from reducing their rate limits to less than a 1/10th within 67 seconds.

But what if the application is a chunk-based video player (e.g., HLS)? If each chunk is 10 seconds or longer, they need more than 6.7 seconds to adopt. IIUC, we allow network elements set their SCONE rate limits equal to what is indicated by congestion signals.

And separately, we should also think about the effects of SCONE packets getting lost.

These points leads me to to argue that the best solution here is to change the misuse detection logic inside network elements.

> I don't know how to implement:
> 
>> IMO, the correct approach for network elements is to retain old limits when the rate limit is reduced, but apply the new rate limits immediately when they are increased.

One possible approach is to require network elements to retain the following two objects:
1. A sliding window that can used to determine total traffic within the last 67 seconds.
2. A three- or four-entry FIFO queue that remembers the most recent rate limits being sent. As rate limits are updated no more than every 20-30 seconds, three- or four-entry queue is sufficient to remember all the rate limits being sent within 67 seconds.

Network elements never reset the sliding window (1). They simply check the conformance by seeing if total traffic within the last 67 seconds is no greater the _maximum_ rate limit retained in the FIFO queue (2) multiplied by 67.

  discard all rate monitoring state,
  and set the last update time to the current time.

A network element that reduces the throughput advice it provides
MUST also discard any state it maintains
about the use of the network under previous, higher rates.
This is necessary to allow senders to discard state
when advice is reduced (see {{algorithm}}).
This avoids cases where an application that has planned usage
might otherwise be completely unable to send due to a lower limits.

A network element that reduces its throughput advice
might also need to add a small delay to allow applications time
to receive and act on the updated advice.

## Flows That Exceed Throughput Advice {#policing}

Network elements that provide throughput advice
can monitor flows --
or sets of flows that are subject to the same throughput limit --
for adherance to that advice.

In the event that a flow exceeds these limits,
a network element could immediately start
enforcing adherence to the advice as though it were a hard limit.
However, this risks creating a situation where communication ceases entirely
for a significant period of time;
that is, up to the period defined in {{time}}.

> A better approach is to disregard any data that was transmitted
> before engaging any hard limits to throughput.

Author: ```suggestion
A better approach is to exclude any data transmitted
before sending reduced throughput advice
from any monitoring that uses the reduced rate.
```
